tools/pipeline-witness/Azure.Sdk.Tools.PipelineWitness/RunProcessor.cs

```diff
@@ -0,0 +1,243 @@
+            var encodedUsernameAndPassword = Convert.ToBase64String(usernameAndPasswordBytes);
+            var header = new AuthenticationHeaderValue("Basic", encodedUsernameAndPassword);
+            return header;
+        }
+        private bool IsValidAzureDevOpsUri(Uri uri)
+        {
+            // We want to throw if we start getting messages from other Azure DevOps
+            // organizations since currently Pipeline Witness is designed to only work
+            // against our instance.
+            return uri.Host == "dev.azure.com" && uri.PathAndQuery.StartsWith("/azure-sdk/");
+        }
+
+        private Guid GetCollectionIdFromProject(TeamProject project)
+        {
+            var url = ((ReferenceLink)project.Links.Links["web"]).Href;
+            var collectionIdAsString = url.Split("/").Last();
+            var collectionId = Guid.Parse(collectionIdAsString);
+            return collectionId;
+        }
+
+        public async Task ProcessRunAsync(Uri runUri)
+        {
+            try
+            {
+                if (!IsValidAzureDevOpsUri(runUri))
+                {
+                    throw new ArgumentOutOfRangeException("Run URI does not point to the Azure SDK instance of Azure DevOps");
+                }
+
+                var runUriPath = runUri.AbsolutePath;
+                var runUriPathSegments = runUri.PathAndQuery.Split("/");
+                var organization = runUriPathSegments[1];
+                var projectGuid = Guid.Parse(runUriPathSegments[2]);
+                var pipelineId = int.Parse(runUriPathSegments[5]);
+                var runId = int.Parse(runUriPathSegments[7]);
+
+                var azureDevOpsPersonalAccessToken = await GetAzureDevOpsPersonalAccessTokenAsync();
+                var credentials = new VssBasicCredential("nobody", azureDevOpsPersonalAccessToken);
+
+                var baseUri = new Uri($"https://dev.azure.com/{organization}");
+                var connection = new VssConnection(baseUri, credentials);
+
+                var buildClient = connection.GetClient<BuildHttpClient>();
+                var projectClient = connection.GetClient<ProjectHttpClient>();
+                var organizationClient = connection.GetClient<OrganizationHttpClient>();
+
+                var build = await buildClient.GetBuildAsync(projectGuid, runId);
+                var project = await projectClient.GetProject(projectGuid.ToString());
+                var pipeline = await buildClient.GetDefinitionAsync(project.Id, build.Definition.Id);
+                var timeline = await buildClient.GetBuildTimelineAsync(projectGuid, runId);
+
+                double agentDurationInSeconds = 0;
+                double queueDurationInSeconds = 0;
+
+                if (timeline != null)
+                {
+                    agentDurationInSeconds = (from record in timeline.Records
+                                              where record.RecordType == "Task"
```

Comment:
Did you find more accuracy at the Task level? I used the Jobs https://dev.azure.com/azure-sdk/internal/_git/azure-sdk-devops?path=%2Fbuild-queues.ps1&version=GBmaster&line=66&lineEnd=67&lineStartColumn=1&lineEndColumn=1&lineStyle=plain, but if the initialize and finalize tasks are accurate maybe that is better.

Reply:
I ended up doing this because I was using the delta between agent start time and job start time to calculate queue time. I don't know for sure, but I do think that this approach excludes as many externalities as possible.